src/msg/async/ProtocolV2.cc

```diff
@@ -49,8 +49,10 @@ void ProtocolV2::run_continuation(CtPtr continuation) {
 }
 
 #define WRITE(B, D, C) write(D, CONTINUATION(C), B)
+#define WRITE_PREAUTH(B, D, C) write(D, CONTINUATION(C), B, true)
 
 #define READ(L, C) read(CONTINUATION(C), L)
+#define READ_PREAUTH(L, C) read(CONTINUATION(C), L, nullptr, true)
```

Comment:
I see following possibilities here:
* introduce extra piece of state: *pre-auth mode*. It would steer the gathering behavior of single, universal `READ`. `ProtocolV2` starts with this flag turned on and drops it in similar places where we do set `session_stream_handlers`. This allows to easily record everything, including both preamble and epilogue.
* Duplicate the `::read_frame()`/`::read_frame_segment()` machinery with e.g. `::read_frame_preauth()`/`::read_frame_segment_preauth()` calling the new `READ_PREAUTH`. Things like `::handle_auth_reply_more()` would do: 
```cpp
return WRITE(more_reply, "auth request more", read_frame_preauth);
 ```
* Capture at the level of frame payload (no `epilogue` nor `preamble`). This can be done by e.g. `::handle_frame_payload()` when it sees a frame that is specific to pre-auth exchange.

Reply:
I'd rather not capture at the frame payload layer.. it's probably mostly sufficient but opens up the possibility of more subtle attacks.

I think a simple ``bool pre_auth = true`` is the simplest way to go!